assayist/client/query.py

```diff
@@ -44,3 +46,83 @@ def get_container_content_sources(build_id):
+    results, _ = neomodel.db.cypher_query(query)
+    rv = []
+    for result in results:
+        rv.append(dict(result[0]))
+    return rv
+
+
+def get_container_built_with_sources(source_locations):
+    """
+    Match container builds that used the input source locations to build the content in the image.
+
+    :param list source_locations: a list of source location dictionaries to match against
+    :return: a list of affected container build Koji IDs
+    :rtype: list
+    """
+    if not source_locations or not isinstance(source_locations, collections.Iterable):
+        raise InvalidInput('The input must be a list of source locations')
+
+    query = """
+        // First get all the input source locations
+        MATCH (input_sl:SourceLocation) WHERE input_sl.url IN [{0}]
+        // Then find all the source locations that are upstream or downstream of the input source
+        // locations recursively. The resulting `sl` variable has all the input source locations
+        // and all the upstream or downstream source locations of the input source locations.
+        MATCH (input_sl)-[:UPSTREAM*0..]-(sl:SourceLocation)
+        // Then find all the source locations that embed the source locations in `sl` recursively.
+        // The resulting `sl` variable will have the contents of the `sl` variable previously and
+        // all the source locations that eventually embed those source locations.
+        MATCH (input_sl)<-[:EMBEDS*0..]-(sl:SourceLocation)
+        // Find all the container images that embed an artifact built from any of the source
+        // locations
+        OPTIONAL MATCH (sl)<-[:BUILT_FROM]-(:Build)-[:PRODUCED]->(:Artifact)<-[:EMBEDS]
+            -(with_embedded_artifact:Artifact {{type: 'container'}})
+        // Find all the container images that were built from any of the source locations
+        OPTIONAL MATCH (sl)<-[:BUILT_FROM]-(:Build {{type: 'container'}})-[:PRODUCED]
+            ->(built_from_sl:Artifact {{type: 'container'}})
+        // Combine the last two matches and store it in a collection called affected_containers
+        WITH COLLECT(with_embedded_artifact) + COLLECT(built_from_sl) AS affected_containers
+        // Unwind the collection so that further queries can take place, and make the values unique
+        UNWIND affected_containers as affected_container
+        WITH DISTINCT affected_container as affected_container
+        // Find all the container image builds that produced containers that were built with any of
+        // the affected containers
+        OPTIONAL MATCH (affected_container)<-[:BUILT_WITH]-(:Artifact {{type: 'container'}})
+            <-[:PRODUCED]-(affected_build:Build)
+        // Find all the container image builds that embed an artifact that was built with an
+        // artifact that was built using the input source locations. This is repeating some of the
+        // query above, but the UNWIND cuts the list down so it seems necessary.
```

Comment:
I don't really understand this comment, and repeating the logic looks error-prone.

Reply:
So if I change the portion above that is repeated `(sl)<-[:BUILT_FROM]-(:Build)-[:PRODUCED]->(:Artifact)` to `(sl)<-[:BUILT_FROM]-(:Build)-[:PRODUCED]->(sl_artifact:Artifact)` and make it a separate `OPTIONAL MATCH`, the `sl_artifact` variable has the correct values for both queries. However, when I do the `UNWIND`, it seems to cut down the values in `sl_artifact`. I'll try rearranging this to see if I can reuse the logic.